src/Client.ts

```diff
@@ -531,7 +537,36 @@ export class Client<
 	): Promise<Query<TDocument>> {
 		const url = await this.buildQueryURL(params)
 
-		return await this.fetch<Query<TDocument>>(url, params)
+		try {
+			return await this.fetch<Query<TDocument>>(url, params)
+		} catch (error) {
+			if (!(error instanceof RefNotFoundError)) {
+				throw error
+			}
+
+			const latestRef = error.message.match(/Master ref is: (?<ref>.*)$/)
+				?.groups?.ref
+			if (!latestRef) {
+				throw error
+			}
+
+			if (this.#retriedRefs.has(latestRef)) {
+				// We only allow retrying a ref once. We'll
+				// likely get the same response on future
+				// retries and don't want to get stuck in a
+				// loop.
+				throw error
+			}
+
+			this.#retriedRefs.add(latestRef)
```

Comment:
I'm afraid this could create some concurrency issues when we dedupe requests:

- T0
  - Page Foo fetches nav document
  - Page Bar fetches nav document
- T1
  - Requests get deduped (only one is made) and fail at the same time with the same error
- T2
  - Page Foo retries the request with the latest ref
  - Page Bar cannot retries the request with the latest ref because it's already being retried by Foo
  - Request cannot be deduped

...or when multiple requests (e.g. `Promise.all(...)`) fail at the same time, which I believe is symptomatic of this kind of error(?)

Maybe we should base this off whether of not a `ref` parameter was provided to the `get` function? i.e. no `ref`: retries with an explicit ref, `ref`: cannot retry

Reply:
You're right; it would likely cause some issues when queries are deduplicated. I removed the `retriedRefs` array and check and simply trust that the provided refs won't cause a loop.

Here's the comment I sent to you [in Slack](https://prismic-team.slack.com/archives/C014VAACCQL/p1730251254283029):

> I removed the logic that prevents the same ref from being retried more than once. You are right about concurrency issues since the retry logic is in the query method, not in the deduplicated fetch job. I'd like to keep the retry in the method so it is close to its intended usage.
>
>I'm wondering if we should still include some kind of loop prevention. I refactored the implementation you reviewed yesterday such that we now have a private `_get()` method that preforms the retry. Since we have that now, it's easy for us to track how many retries and halt if we get too deep.
>
>I think we should include some kind of loop protection to be safe, but I wanted to get your thoughts first.

---

> Maybe we should just retry directly instead of recursively calling the function after that?

I'd rather not retry directly since it means we can only follow the API's response once. If a website has a very stale cache with multiple chained invalid ref responses, we should be able to use each of them.

We log a warning each time we retry, so devs will still be nudged to fix their caching issue.

---

> Or base the retry off whether or not we had an explicit `ref` parameter(?)

A retry may be necessary whether or not an explicit `ref` is provided. Websites may have a stale `/api/v2`, which would result in a stale master ref being used implicitly.

---

> 💡 #idea: Maybe we should clear the repository cache when this happens so that other requests (during the cache's 5s lifespan) can fetch a fresh ref?

Maybe! This fix is really intended for Next.js' indefinite cache, which works outside the client's internal 5s cache.

I can't think of any downside to clearing the internal cache. It would resolve the tiny chance that a burst of requests from the same client results uses a stale, internally cached ref. If you can't think of any downsides, I'll add it in.

Reply:
Thanks for the clarification!

- I agree we should have some kind of loop prevention yes ♻️ 
- I'm aligned with clearing the cache, it's mostly free and will clear some edge cases 👌